Connection: call verify_secrets() in verify_action()

kbrown9 opened the pull request:
The method `Manager::verify_secrets()` is the basically the same as the method `Jetpack_XMLRPC_Server::verify_action()`. Remove the duplication of code by calling
`Manager::verify_secrets()` from `Jetpack_XMLRPC_Server::verify_action()`.

#### Changes proposed in this Pull Request:
* Call `Manager::verify_secrets()` from `Jetpack_XMLRPC_Server::verify_action()`.
* Make `Manager::verify_secrets()` public.
* Fix the error code in the `elseif ( ! ctype_digit( (string) $user_id )` branch of `Manager::verify_secrets()`. It should be "state_malformed".

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
* This improves an existing feature.

#### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* see [this comment](https://github.com/Automattic/jetpack/pull/13886#pullrequestreview-310180626) and [this comment](https://github.com/Automattic/jetpack/pull/13886#issuecomment-548839119)

#### Proposed changelog entry for your changes:
<!-- Please do not leave this empty. If no changelog entry needed, state as such. -->
* Not needed.

kbrown9 committed:
[not verified] The method Manager::verify_secrets() is the basically the same as the method
Jetpack_XMLRPC_Server::verify_action(). Remove the duplication of code by calling
Manager::verify_secrets() from Jetpack_XMLRPC_Server::verify_action().

zinigor committed:
[not verified] Turned return statements into assignments to make CodeClimate happy.

zinigor commented:
I have added tracking back to the code, sorry for the ping, Paul :)

The diff got a bit confusing, but essentially what I did was add an action handler per every begin, success, and failure actions. Now that I did it, I realize that we still have tracking in the connection package, do you think we can start moving these out to a separate package that would just add these event handlers?

kbrown9 committed:
Connection: verify_action() needs to return an IXR_Error on failure

The docblock for verify_action() shows that a WP_Error is returned on failure, but this
is incorrect. Any errors generated in verify_action() are passed through the error() method,
which converts them to IXR_Errors. So verify_action() actually returns on IXR_Error on failure.

Make the following changes:
* In verify_action(), check whether verify_secrets() returned an error. If it did, pass that error
  to the error() method to convert it to an IXR_Error.
* Fix the docblock for verify_action().
* Add @return to the verify_secrets() docblock.

kbrown9 committed:
Connection: Check that $params indexes 0 and 1 are set

zinigor reviewed (dismissed):
This seems to be working pretty well. Both register and authorize events work fine, to test them I reconnected the site by deactivating Jetpack (thus deleting all options) and reconnecting it to WordPress.com.
Tracks are trickier to test, but it's possible if you set a logging call somewhere in the client. Plus, the register track only happens if the user has previously agreed to the ToS. I can see that happening if the options are all in place, but the token got broken somehow, and we force the re-registration of the site to fix that.
Currently there's no way to test the publicize track because there's no code left that sends a publicize API request to dotcom that would respond with a secrets check.
Edit: there are still publicize events that can be triggered from Jetpack, at least a deletion can be done using WP CLI.

kbrown9 commented:
> This seems to be working pretty well. Both register and authorize events work fine, to test them I reconnected the site by deactivating Jetpack (thus deleting all options) and reconnecting it to WordPress.com.

Thanks! 

I also tested the case where `verify_action()` returns an error. I did this by altering the secret in `verify_action()` so that it doesn't match the stored secret. Then I deactivated Jetpack and attempted to reconnect. The connection should fail, and it did.

gravityrail commented on packages/connection/src/Manager.php:
```diff
@@ -1290,23 +1292,23 @@ protected function verify_secrets( $action, $secret_1, $user_id ) {
 		}
 
 		if ( self::SECRETS_MISSING === $stored_secrets ) {
```
Because we are now exiting later, rather than returning on the first error, we may end up overwriting the $error value before returning it. For example, if `verify_secret_1_missing` is recorded, then `verify_secrets_expired`, we will only see the later error and not the earlier one. This could make things a little harder to debug in terms of finding the root cause.

I would recommend making this check an elsif like all the ones above it.

kbrown9 replied:
Change made in 1f7a673.

gravityrail commented on packages/connection/src/Manager.php:
```diff
@@ -1334,6 +1336,11 @@ protected function verify_secrets( $action, $secret_1, $user_id ) {
 		}
 
+		// Something went wrong during the checks, returning the error.
+		if ( null !== $error ) {
```
This check doesn't have to be so narrow, you could just use `if ( ! empty( $error ) ) ...`

kbrown9 replied:
Change made in 1f7a673.

gravityrail reviewed (dismissed):
Nice cleanup. Couple small issues.

kbrown9 committed:
Connection: Minor code improvements for verify_secrets

zinigor reviewed (approved):
LGTM!
